crosscluster/logical: avoid error handling race

stevendanna opened the pull request:
Epic: none
Release note: None

stevendanna committed:
crosscluster/logical: avoid error handling race

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/127726)
<!-- Reviewable:end -->

stevendanna commented:
Writing a test here now. Just getting this PR out to put a flag in the ground.

stevendanna committed:
crosscluster: add String() method to Event type

Epic: none
Release note: None

stevendanna committed:
crosscluster/logical: fix another error race

Here, we were occasionally not capturing the error returned from the
subscription.

Here, we assume an error is always coming. This appears to be the case
to me as the only way in non-test code that the subscription returns
and the event channel closes is if there is an error.  But, I've added
a timeout in case that isn't correct.

Epic: none
RElease note: None

stevendanna committed:
crosscluster/logical: add random stream test

This improves the randomgen to support basic single-table logical
streaming with a single partition and then adds a test using this new
feature.

Under stress, this test catches both of the recently fixed error
handling race conditions.

Epic: none
Release note: None

stevendanna committed:
crosscluster/logical: increase timeout under race

Epic: none
Release note: None

msbutler commented on pkg/ccl/crosscluster/streamclient/randclient/random_stream_client.go:
```diff
@@ -82,6 +82,13 @@ const (
 	IngestionTablePrefix = "foo"
 )
 
+func init() {
```
valiant effort to move the random client to its own package :D

msbutler commented on pkg/ccl/crosscluster/streamclient/randclient/random_stream_client.go:
```diff
@@ -232,23 +248,31 @@ type randomEventGenerator struct {
-	if err != nil {
-		return nil, err
+
+	var systemKVs []roachpb.KeyValue
```
no need to change this now, but it's not clear to me what this systemKVs knob is adding to test coverage.

stevendanna replied:
Not clear to me either. Perhaps the original idea was that these would be needed to allow us to read from the table post promotion. But the tests don't actually do that. And, even if they did, all it would be testing is that the table is readable, there would be nothing to compare it to.

msbutler commented on pkg/ccl/crosscluster/streamclient/randclient/random_stream_client.go:
```diff
@@ -275,33 +298,40 @@ func (r *randomEventGenerator) generateNewEvent() crosscluster.Event {
+		return nil, err
+	}
+	switch eventType {
+	case randomKV:
```
wow this is much cleaner. thank you for doing this!

msbutler commented on pkg/ccl/crosscluster/streamclient/randclient/random_stream_client.go:
```diff
@@ -614,61 +651,65 @@ func (r *randomStreamSubscription) Events() <-chan crosscluster.Event {
+			datums = append(datums, tree.DNull)
+		} else {
+			datums = append(datums,
+				randgen.RandDatum(rng, c.GetType(), c.IsNullable()))
```
!!!!

msbutler commented on pkg/ccl/crosscluster/streamclient/randclient/random_stream_client.go:
```diff
@@ -745,14 +786,49 @@ func (m *RandomStreamClient) ClearInterceptors() {
+	if len(req.TableNames) > 1 {
+		return nil, errors.AssertionFailedf("random stream only supports 1 table")
+	}
+	if m.config.existingTableID == 0 {
```
this assertion should be removed, right?

msbutler commented on pkg/ccl/crosscluster/streamclient/randclient/random_stream_client.go:
```diff
@@ -745,14 +786,49 @@ func (m *RandomStreamClient) ClearInterceptors() {
+	srcCodec := keys.MakeSQLCodec(m.config.tenantID)
+	tableSpan := m.tableDesc.TableSpan(srcCodec)
+	partitionURI := m.config.URL(int(m.tableDesc.GetID()), false, false)
+	topology.Partitions = append(topology.Partitions,
```
should we assert that `m.config.numPartitions` must equal 1?

msbutler commented on pkg/ccl/crosscluster/logical/logical_replication_job_test.go:
```diff
@@ -489,6 +492,62 @@ func TestRandomTables(t *testing.T) {
+		t.Logf("%s", string(allstacks.Get()))
+		t.Fatal("timed out waiting for events")
+	}
+
```
for another day: we could run some validation on the LDR job. E.g: if fewer than numRetryable errors were sent, the stream is still running.

msbutler replied:
one quick validation would be to assert that nothing is in the dlq?

stevendanna replied:
I'm not sure that is viable because there is not guarantee that our random stream doesn't produce conflicting KVs.

msbutler reviewed (commented):
This is awesome! Left a bunch of nits.

stevendanna commented on pkg/ccl/crosscluster/streamclient/randclient/random_stream_client.go:
```diff
@@ -614,61 +651,65 @@ func (r *randomStreamSubscription) Events() <-chan crosscluster.Event {
-		Key:   k,
-		Value: v,
+func randDatumsForTable(rng *rand.Rand, td catalog.TableDescriptor) ([]tree.Datum, error) {
+	columns := td.PublicColumns()
```
Good call. Done.

stevendanna commented:
bors r=msbutler
